src/target/at32f43x.c

```diff
@@ -0,0 +1,376 @@
+}
+
+static inline uint32_t at32f43_bank_offset_for(const target_addr_t addr, const target_addr_t bank_split)
+{
+	if (addr >= bank_split)
+		return FLASH_BANK2_REG_OFFSET;
+	return FLASH_BANK1_REG_OFFSET;
+}
+
+static bool at32f43_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	target_s *target = flash->t;
+	const at32f43_flash_s *const sf = (at32f43_flash_s *)flash;
+
+	/* Erase range begins in bank 1? Unlock bank 1 */
+	if (addr < sf->bank_split && !at32f43_flash_unlock(target, 0))
+		return false;
+	/* Erase range ends in bank 2? Unlock bank 2 */
+	const target_addr_t end = addr + len - 1U;
+	if (end >= sf->bank_split && !at32f43_flash_unlock(target, FLASH_BANK2_REG_OFFSET))
+		return false;
+
+	for (size_t offset = 0; offset < len; offset += flash->blocksize) {
+		const uint32_t bank_reg_offset = at32f43_bank_offset_for(addr + offset, sf->bank_split);
+		at32f43_flash_clear_eop(target, bank_reg_offset);
+
+		/* Prepare for page/sector erase */
+		target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS);
+		/* Select erased sector by its address */
+		target_mem_write32(target, AT32F435_FLASH_ADDR + bank_reg_offset, addr + offset);
+		/* Start sector erase operation */
+		target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS | FLASH_CTRL_ERSTR);
+
+		/* Datasheet: page erase takes 50ms (typ), 500ms (max) */
+		if (!at32f43_flash_busy_wait(target, bank_reg_offset, NULL))
+			return false;
+	}
+	return true;
+}
+
+static inline size_t at32f43_range_in_bank1(const target_addr_t addr, const size_t len, const target_addr_t bank_split)
+{
+	if (addr >= bank_split)
+		return 0;
+	if (addr + len > bank_split)
+		return bank_split - addr;
+	return len;
+}
+
+static bool at32f43_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len)
+{
+	target_s *target = flash->t;
+	const at32f43_flash_s *const sf = (at32f43_flash_s *)flash;
+	const size_t range = at32f43_range_in_bank1(dest, len, sf->bank_split);
+	const align_e psize = ALIGN_WORD;
+
+	/* There is something to write to bank 1 */
+	if (range) {
+		at32f43_flash_clear_eop(target, FLASH_BANK1_REG_OFFSET);
+
+		target_mem_write32(target, AT32F435_FLASH_CTRL, FLASH_CTRL_FPRGM);
+		cortexm_mem_write_sized(target, dest, src, range, psize);
+
+		/* Datasheet: flash programming takes 50us (typ), 200us (max) */
+		if (!at32f43_flash_busy_wait(target, FLASH_BANK1_REG_OFFSET, NULL))
+			return false;
+	}
+
+	/* For dual-bank targets write the remainder to bank 2 */
+	const size_t remainder = len - range;
+	if (sf->bank_split && remainder) {
+		const uint8_t *data = src;
+		at32f43_flash_clear_eop(target, FLASH_BANK2_REG_OFFSET);
+
+		target_mem_write32(target, AT32F435_FLASH_CTRL + FLASH_BANK2_REG_OFFSET, FLASH_CTRL_FPRGM);
+		cortexm_mem_write_sized(target, dest + range, data + range, remainder, psize);
+
+		if (!at32f43_flash_busy_wait(target, FLASH_BANK2_REG_OFFSET, NULL))
+			return false;
+	}
```

Comment:
I don't understand the need for this split / remainder handling, maybe you could clear it up. Aren't the 2 banks handled by 2 different flash structs? if so a call to `flash_write` for bank1 flash should never have any data for bank2 flash.

Reply:
On second thought you're right, a single write call won't cross bank boundary, and that is guaranteed by new target flash API which flushes buffered writes `writesize`-wise. But I still need to dispatch writes to the correct bank (registers) as there's only one routine used with both regions.